compiler/rustc_infer/src/infer/combine.rs

```diff
@@ -119,20 +119,32 @@ impl<'tcx> InferCtxt<'tcx> {
                 self.unify_float_variable(!a_is_expected, v_id, v)
             }
 
-            // All other cases of inference are errors
-            (&ty::Infer(_), _) | (_, &ty::Infer(_)) => {
-                Err(TypeError::Sorts(ty::relate::expected_found(relation, a, b)))
-            }
-
             (ty::Alias(AliasKind::Projection, _), _) if self.tcx.trait_solver_next() => {
+                assert!(!matches!(
+                    b.kind(),
+                    ty::Infer(
+                        ty::TyVar(_) | ty::FreshTy(_) | ty::FreshIntTy(_) | ty::FreshFloatTy(_)
+                    )
+                ));
```

Comment:
instead of `assert!(!matches!`, can this be a `match { infer... => bug!(..), _ => { relation.register_type_equate_obligation(a, b); Ok(b) }` in both branches? This is fairly hard to read imo

Reply:
Actually, we don't need two branches anyways. I tried reworking the match, though it still may be kind of ugly.